### ISSUE TITLE
What are the expectations for rendering Scribe based content?

### ISSUE THREAD

**Issue by gmoreira** [madebymany/sir-trevor-js#314]
I am a little confused about the direction of this project with the move from Markdown to Scribe and well-formed HTML. I can only assume a lot of this has already been discussed offline. But what I want to understand is, what is the expectation now for the display of this structured data (where ever it may appear).

I have reviewed #301 and #274 and I understand what was done. I have also just submitted #313 to fix the List block in a similar way to #301 and I imagine the same needs to be done for other blocks like Headings?

So where my confusion lies is in the rendering of block data. If we look at the default Heading block, a simple heading like "This is a heading" will generate the following ST block data:

`{"data":[{"type":"heading","data":{"text":"<p>This is a heading<br></p>"}}]}`

If we look at rendering this content content on the web then my questions is: what is the expectation?

My initial though is to render the text property data within a H1 element like so:

`<h1><p>This is a heading<br></p></h1>`

However doesn't this create invalid HTML? We've got `<p>` tags within `<h1>` tags which doesn't seem correct.

Or is this exactly the expectation for "end consumers/renderers" of ST block data?

**Comment by higgis** [comment]
Hi @gmoreira,

Thanks for your thoughtful questions.

Hopefully I can provide a summary of the thinking behind moving away from Markdown and then offer some specific remarks about the heading block (and other blocks). Apologies that the reasoning behind the decision to make this move was not more public or well-documented.

### Why we moved away from Markdown

There are two reasons for moving away from Markdown as a storage format, one pragmatic and one philosophical. 

The pragmatic one is that Markdown is not easy to parse correctly, and the RegExes that did that job for us were becoming very difficult to maintain. We could have refactored the way we did this or used a different approach entirely, but ultimately it led us to think quite deeply about why we were using Markdown at all, and this led us to the philosophical decision.

We think that Markdown is great for some things, the primary one being making it relatively easy for non technical people to author structured text and have that text converted into an appropriate representation (such as HTML) automatically. However the editing interface in Sir Trevor does not accept Markdown as an input, so this benefit is lost.

The only sensible solution to allowing non-technical users to add structured text in a graphically rich way using a browser is contenteditable at the time of writing and for the foreseeable future. As we all know, contenteditable deals in HTML. A limited subset of HTML is also a good choice for structured content storage; it's at least as good as Markdown anyway (in my opinion a lot better). There is no obligation to present the data as HTML once it has been retrieved from storage, but many many parsers exist to allow you to reason about the contents of a string containing HTML, on all platforms and in all languages. 

If you want to present the structured text as HTML, then you have next to no work to do. If you want to do something non-HTML based you have a small amount of work to do, but then you do if it's stored as Markdown anyway, so this is no imposition.

The issue was that we were allowing users to input structured text using contenteditable, converting that (rather badly) to Markdown, storing that and then converting it back to HTML or some other format for presentation. What's slightly perverse is that at no stage is Markdown used except as an intermediate format. We do not know about anyone using it to present Markdown at the end of the process. (A markdown block which let you actually write in that format would be a good idea and would avoid the need to convert it to and from some other format for editing.)

Thus, Markdown is in effect an arbitrary choice of storage format given that it is less expressive than HTML and, although very widely supported, less easy to use at the point of presentation.

Finally, we have never been happy with our contenteditable wrapper and Scribe does a much, much better job of this.

For all of these reasons, we decided to move away from Markdown as a storage format for structured text.

### How other blocks should work

In my view, it is incorrect to store the contents of highly specific blocks like the heading block using HTML inside the JSON. I don't see what is gained: the effort of rendering plain text contents of it inside an ```<h{n}>``` tag is tiny if you're rendering it for a browser, and obviously there may be completely different ways you might want to treat a heading  (such as in an iOS app) which would require you to strip any HTML out of the block first. 

i.e:

```<h1><%= block.text %></h1>```

for a browser, or:

```heading.text = block.text;```

in, say, iOS. Otherwise the iOS code has to become:

```heading.text = [HtmlHelper stripHtml:block.text];```

or something equally ugly.

We should be prioritising two things above all others - ease of use for editors, and programmer happiness (in that order). I don't see how storing HTML in the block data advances either of these causes. As you rightly point out, it's also not valid HTML which makes things considerably worse.

For me, the block data for a heading should look something like this:

```{"data":[{"type":"heading","data":{"text":"This is a heading","level": 1}}]}```

Just because HTML supports headings as a first class concept and the editing interface is being hosted in a browser does not in any way justify using HTML as a storage mechanism.

However, I think there's a strong case for storing lists in HTML, since there's a good construct for that and it makes it easy to present again after retrieval from storage. It also makes it easy to present a usable editing interface with a minimum of effort. I would definitely avoid Markdown for lists because nesting is so much less expressive in Markdown than it is in HTML.

It's probably less easy to use for non-browser presentation, but that's still the case with Markdown and there are the nesting issues that I mentioned to take account of there too. On balance, therefore, I think that HTML is the correct format for lists, while remaining fairly far from ideal.

In other words, I believe that it's a decision that should be made on a case by case basis for each block. 

### Summary

The change to HTML for structured free form text had been carefully considered, and we believe that it is the correct decision. We would be very interested to hear counter arguments, but they'd need to be very strong at this stage to make us consider changing our minds.

There is still a fair bit of detailed work to do to clean up other blocks (such as heading) and make them behave more sensibly. In my view the current implementation that you detailed is unacceptable and should be fixed as soon as possible. 

Other blocks will have different tradeoff and these need to be considered on a case by case basis.

I hope that helps.

James

**Comment by zgryw** [comment]
If it doesn't answer all your questions feel free to reopen :+1:
